### DISCUSSION TITLE
developers/fxplug/

### DISCUSSION THREAD

**Discussion by giscus** [CommandPost/FCPCafe#306] [Website Discussions]
# developers/fxplug/

This section of the site is dedicated to FxPlug.

https://fcp.cafe/developers/fxplug/

**Comment by mkuhnkey** [comment]
Service.pluginkit/Contents/MacOS/../Frameworks/PluginManager.framework/Versions/B/PluginManager' (fat file, but missing compatible architecture (have 'i386,x86_64', need 'arm64'))

MacBook Pro (16 inch) M1 (arm64e) maOS Sonoma with Xcode 15 does not support FxPlug PluginManager's "fat-binary"

**Reply by latenitefilms** [reply]
I’m not seeing any issues here? Are you using latest FxPlug?

**Reply by mkuhnkey** [reply]
Yes. FxPlug 4.2.x
What is your "lido" result for FxPlug PluginManager?
I do not use rosette...M1 silicon arm64e
I assume you have built FxPlug plugin that interfaces with Motion5 via XPC?

**Reply by latenitefilms** [reply]
I'm on a 2021 16-inch MacBook Pro (M1 Max, 64GB RAM, 8TB SSD) running Final Cut Pro 10.7, Sonoma 14.1.2 and Xcode 15.1.

Yes, I have multiple FxPlug's - see: https://fcp.cafe/latenite/

I'm currently building a new one in Swift: https://metaburner.pro

**Reply by mkuhnkey** [reply]
I have tried FxPlug from Xcode Template (Swift or Objective-C)...and followed Development documents for both Xcode Template and Manual install...builds are successful and attach to Motion, but PluginManager fails with "fat-binary" with no asm64 architecture.

**Reply by latenitefilms** [reply]
Very strange. Maybe try download **FxPlug v4.2.9** again?

https://developer.apple.com/download/all/?q=FxPlug

**Reply by mkuhnkey** [reply]
I've broken something....
 % lipo PluginManager -info
fatal error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: can't open input file: PluginManager (No such file or directory)

Did not get this error yesterday.....back to the rabbit hole.  But I am curious what you get with "% lipo PluginManager -info"

Thank-You...I'm impressed by your work!

**Reply by latenitefilms** [reply]
Can you share the full Terminal command I need to enter?

**Reply by mkuhnkey** [reply]
That's it...
% lipo PluginManager -info <RETURN>

Should provide the "Universal Binary" aka "fat-binary" with the supported architectures.

**Reply by latenitefilms** [reply]
`fatal error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: can't open input file: PluginManager (No such file or directory)`

**Reply by mkuhnkey** [reply]
....I'll be in the corner for awhile...banging my head on the wall. Thanks for the "info" and your time,  much appreciated!

**Reply by mkuhnkey** [reply]
Followup....Resolved by deleting "Derived Data" in Xcode.  FxPlug4 Good to Go!

**Comment by mkuhnkey** [comment]
Examples...FxPlug4.2.9 Have me "banging-my-head-in-the-corner"...Specifically "FxPictureInPicture" (macOS Sonoma M1, Xcode 15).....dyld[88289]: Library not loaded: @rpath/PluginManager.framework/Versions/B/PluginManager
  Referenced from: <60A0B8E5-7C2F-304E-B5FB-BE60279C31A9> /Users/mikekuhnkey/Library/Developer/Xcode/DerivedData/FxPictureInPicture-fcpuksxzjnmgkqbijshcugpgtthp/Build/Products/Release/FxPictureInPicture.app/Contents/PlugIns/FxPictureInPicture.pluginkit/Contents/Frameworks/FxPlug.framework/Versions/A/FxPlug .......Note "Versions/A" is supported by FxPlux4 (x86_64 & arm64).......However PluginManager is limited to "Version/B" (i386 & 86_64)?..........Resulting in
"missing compatible architecture (have 'i386,x86_64', need 'arm64'".....PluginManager.tbd <--tapi is a problem??

**Reply by latenitefilms** [reply]
The `FxPictureInPicture` sample project in FxPlug `v4.2.9` works fine for me out-of-the-box?

I'm on a 16-inch MacBook Pro (M1 Max, 64GB RAM, 8TB SSD) running macOS Sonoma 14.1.2 and Xcode 15.1.

**Comment by mkuhnkey** [comment]
Did you run a "Release" build with no Debug?  Was it working in Motion/FCPx?  I ran "out-of-the-box" also...Deleted "DerivedData" and "Clean Build"...exact same issue with PluginManager missing arm64. Was able to build FxBrightness attached to Motion but it didn't produce an "effect" in Templates.

**Reply by latenitefilms** [reply]
Yes, it works fine in Motion:

<img width="453" alt="image" src="https://github.com/CommandPost/FCPCafe/assets/22286696/c0d8a8f4-68fc-441d-bcad-6e1d4cffbe0a">

**Comment by mkuhnkey** [comment]
I'm also getting "Thread 1 signal SIGBRT"....assume it's related PluginManager (Xcode error)

**Comment by mkuhnkey** [comment]
Was able to "manually create" FxPlug4 in Xcode (Developer Documentation)...but it used Swift in the document.  Was able to attach to Motion, but seemed to hang "ProRegistration" and the Motion icon just bounced regularly.  So,some success with FXPlug4 and Swift but the "Examples" FxPlug4.2..9 are low-level Objective-C for FxPictureInPicture.

**Reply by latenitefilms** [reply]
When you create a new FxPlug4 using the Xcode template, you can select Objective-C OR Swift.

**Comment by mkuhnkey** [comment]
Tried BOTH...Objective-C used XIB for interface....and Swift was the other.

I'm impressed by your results, but really frustrated!  Did "EXAMPLES" Subcategory install by itself in Motion?

**Reply by latenitefilms** [reply]
Yes. When you build the application in Xcode and run it, macOS detects the FxPlug4 extension inside the application bundle, and registers it. Once registered Motion adds it to the Library, based on the settings you have in the FxPlug's `plist` file.

**Comment by mkuhnkey** [comment]
Any Thoughts....
#0	0x0000000187f6255c in __abort_with_payload ()
#1	0x0000000187f6eb10 in abort_with_payload_wrapper_internal ()
#2	0x0000000187f6eb44 in abort_with_payload ()
#3	0x0000000187ef5584 in dyld4::halt(char const*, dyld4::StructuredError const*) ()
#4	0x0000000187ef2254 in dyld4::prepare(dyld4::APIs&, dyld3::MachOAnalyzer const*) ()
#5	0x0000000187ef0edc in start ()
Guess I'll go look at "Wrapper"....

**Reply by latenitefilms** [reply]
I'm really not sure where you're going wrong... Maybe try...

1) Uninstall Xcode
2) Restart Computer
3) Reinstall Xcode
4) Download & Install latest FxPlug4 SDK
5) Copy the FxPlug4 SDK Examples to your desktop
6) Open one of the sample projects, try and build it, then open in Motion and see if it appears

**Reply by mkuhnkey** [reply]
Deleted everything....but not Xcode..ran 1 Example...it built and scheme was set to Motion. It started, and displayed Motion...no EXAMPLES.  Ran ALL the EXAMPLES....now each one failed with the PluginManager issue!  So, yessir taking your advice!

**Reply by latenitefilms** [reply]
The scheme should be set to the Wrapper Application by default?

<img width="606" alt="image" src="https://github.com/CommandPost/FCPCafe/assets/22286696/d2cfbf4f-a2d1-432a-9ba5-494f5c6f3bc9">

**Reply by mkuhnkey** [reply]
Did you get Warnings about "Trust and still load" for Examples?....and NoJoy....on clean slate 🤦‍♂️

**Reply by latenitefilms** [reply]
No warnings or errors. Just works straight out of the box.

**Reply by latenitefilms** [reply]
Here's what happens when I build the project:

<img width="1429" alt="image" src="https://github.com/CommandPost/FCPCafe/assets/22286696/2032cd8a-0aeb-4c98-a7dd-91fdd853f570">

No changes or settings changed - just literally opening the FxPlug4 sample project.

I can jump on Zoom with you next week to problem solve?

**Reply by mkuhnkey** [reply]
All that works fine when scheme is set to FXPictureInPicture.app...It's when I point scheme to Motion?  Am I doing something stupid with the scheme settings?

**Reply by latenitefilms** [reply]
You're misunderstanding... you need to build the Wrapper Application. Once the Wrapper Application runs, behind the scenes macOS will automatically register the plugin, and when you start Motion it will be there.

**Reply by mkuhnkey** [reply]
Yep! That's what's not happening...stupid scheme settings...Fixed...But when I run the FxPictureInPicture with scheme set to "FxPictureInPicture.app" I get the object displayed...on/off...still no Motion "filter" when  Motion is launched?

**Reply by mkuhnkey** [reply]
Good information! I'll clean up a bit, build and such...Give it a rest and try tomorrow.

**Reply by mkuhnkey** [reply]
...for some reason "pluginkit -m FxPlug" is not working now?

**Reply by mkuhnkey** [reply]
pluginkit is working...debug of Motion now returns:.........Unable to load PlugIn: FxPictureInPicture due to error: 4, Extension process exited while/after acquiring startup assertion.

**Reply by mkuhnkey** [reply]
I've broken the "out-of-process" capability in XPC. Motion starts and lives.  History..Re-installed Xcode and installed FxPlug4.sdk...all FxPlug EXAMPLES showed up in the Xcode "browser" when started initially...bit of troubleshooting required. Seems to be issue with Xcode "build phases" while working with FxPlug4 (Copy and Code Sign PluginManager.framework) under XPC Service.  Xcode --> Debug Executable -->Motion (using "Console" in MacOS) confirms crash of "out-of-process" capability.

**Reply by mkuhnkey** [reply]
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               FxBrightness XPC Service [30943]
Path:                  /Users/USER/Library/Developer/Xcode/DerivedData/FxBrightness-emwaebtuzspagudpnqzkpldhxcmq/Build/Products/Debug/FxBrightness.app/Contents/PlugIns/FxBrightness XPC Service.pluginkit/Contents/MacOS/FxBrightness XPC Service
Identifier:            FxBrightness-XPC-Service
Version:               1.0 (1)
Code Type:             ARM-64 (Native)
Parent Process:        launchd [1]
Responsible:           Motion [30940]
User ID:               501

Date/Time:             2024-01-07 08:43:15.7668 -0600
OS Version:            macOS 14.2.1 (23C71)
Report Version:        12
Anonymous UUID:        D2F7717B-4DF4-A459-8F86-25BC01D39092

Sleep/Wake UUID:       024E772F-75AD-4B7B-AF92-03CBC2BED272

Time Awake Since Boot: 110000 seconds
Time Since Wake:       7120 seconds

System Integrity Protection: enabled

Crashed Thread:        0

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Termination Reason:    Namespace DYLD, Code 1 Library missing
Library not loaded: @rpath/PluginManager.framework/Versions/B/PluginManager
Referenced from: <60A0B8E5-7C2F-304E-B5FB-BE60279C31A9> /Users/USER/Library/Developer/Xcode/DerivedData/FxBrightness-emwaebtuzspagudpnqzkpldhxcmq/Build/Products/Debug/FxBrightness.app/Contents/PlugIns/FxBrightness XPC Service.pluginkit/Contents/Frameworks/FxPlug.framework/Versions/A/FxPlug
Reason: tried: '/Users/mikekuhnkey/Library/Developer/Xcode/DerivedData/FxBrightness-emwaebtuzspagudpnqzkpldhxcmq/Build/Products/Debug/FxBrightness.app/Contents/PlugIns/FxBrightness XPC Service.pluginkit/Contents/Frameworks/FxPlug.framework/Versions/A/Frameworks/PluginManager.framework/Versions/B/PluginManager' (no such file), '/System/Library/PrivateFrameworks/PluginManager.framework/Versions/B/PluginManager' (no such file, not in dyld cache), '/System/Volumes/Preboot/Cryptexes/OS/System/Library/PrivateFrameworks/PluginManager.framework/Versions/B/PluginManager' (no such file), '/usr/lib/swift/PluginManager.framework/Versions/B/PluginMan
(terminated at launch; ignore backtrace)

**Comment by mkuhnkey** [comment]
**Think this is "rsync" issue:** (architecture diffs between PluginManager & FxPlug)
Appears to be why DYLD doesn't happen?
mkdir -p "${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}" <~~~ still puzzles me
% file /Library/Developer/Frameworks/PluginManager.framework/Versions/B/PluginManager
/Library/Developer/Frameworks/PluginManager.framework/Versions/B/PluginManager: Mach-O universal binary with 2 architectures: [i386:Mach-O dynamically linked shared library i386] [x86_64:Mach-O 64-bit dynamically linked shared library x86_64]

rsync --archive --links --whole-file --no-owner --no-group --checksum --exclude='Headers' "/Library/Developer/Frameworks/PluginManager.framework/" "${TARGET_BUILD_DIR}/${FRAMEWORKS_FOLDER_PATH}/PluginManager.framework/"
if [ "$?" -ne 0 ]; then echo "error: rsync failed; couldn't embed PluginManager.framework"; exit 1; fi

%file /Library/Developer/Frameworks/FxPlug.framework/Versions/A/FxPlug
/Library/Developer/Frameworks/FxPlug.framework/Versions/A/FxPlug: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]

**Comment by mkuhnkey** [comment]
Terminal:
<img width="1117" alt="FxPictureInPicture" src="https://github.com/CommandPost/FCPCafe/assets/17168164/88b42cb0-d49f-4698-9995-584a0b7d18c2">

**Reply by mkuhnkey** [reply]
FxPlug returns:
 % file /Library/Developer/Frameworks/FxPlug.framework/Versions/A/Fxplug
/Library/Developer/Frameworks/FxPlug.framework/Versions/A/Fxplug: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/Library/Developer/Frameworks/FxPlug.framework/Versions/A/Fxplug (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
/Library/Developer/Frameworks/FxPlug.framework/Versions/A/Fxplug (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64
